tools/cassandra/handler.go

```diff
@@ -47,24 +48,48 @@ type SetupSchemaConfig struct {
 func VerifyCompatibleVersion(
 	cfg config.Persistence,
 ) error {
-
-	ds, ok := cfg.DataStores[cfg.DefaultStore]
-	if ok && ds.Cassandra != nil {
-		err := CheckCompatibleVersion(*ds.Cassandra, cassandra.Version)
-		if err != nil {
+	if ds, ok := cfg.DataStores[cfg.DefaultStore]; ok {
+		if err := verifyCompatibleVersion(ds, map[string]string{
+			cassandra_db.PluginName: cassandra.Version,
+		}); err != nil {
 			return err
 		}
 	}
-	ds, ok = cfg.DataStores[cfg.VisibilityStore]
-	if ok && ds.Cassandra != nil {
-		err := CheckCompatibleVersion(*ds.Cassandra, cassandra.VisibilityVersion)
-		if err != nil {
+
+	if ds, ok := cfg.DataStores[cfg.VisibilityStore]; ok {
+		if err := verifyCompatibleVersion(ds, map[string]string{
+			cassandra_db.PluginName: cassandra.VisibilityVersion,
+		}); err != nil {
 			return err
 		}
 	}
+
 	return nil
 }
 
+func verifyCompatibleVersion(
+	ds config.DataStore,
+	expectedVersions map[string]string,
+) error {
+	var expectedVersion string
+	var cfg config.NoSQL
+	if ds.NoSQL != nil {
+		var ok bool
+		if expectedVersion, ok = expectedVersions[ds.NoSQL.PluginName]; !ok {
+			return fmt.Errorf("unknown NoSQL plugin name: %v", ds.NoSQL.PluginName)
+		}
+		cfg = *ds.NoSQL
+	} else if ds.Cassandra != nil {
+		expectedVersion = cassandra.VisibilityVersion
+		cfg = *ds.Cassandra
+	} else {
+		// not using nosql or cassandra
```

Comment:
Nit: returning error here should be fine.

Reply:
I can't return error here, otherwise server using sql config can't start as we always execute the verification function for both cassandra and sql. See: https://github.com/uber/cadence/blob/master/cmd/server/cadence/cadence.go#L65